sympy/combinatorics/rewritingsystem.py

```diff
@@ -285,3 +292,132 @@ def reduce(self, word, exclude=None):
+                fsm.add_state(rule, is_dead=True)
+
+        # Add accept states.
+        for r in all_rules:
+            prop_prefix = proper_prefixes[r]
+            for elem in prop_prefix:
+                if not elem in fsm.states:
+                    fsm.add_state(elem, is_accept=True)
+
+        # Add transitions for every state
+        for state in fsm.states:
+            current_state_name = state
+            current_state_type = fsm.states[state].state_type
+            if current_state_type == "start":
+                for letter in automaton_alphabet:
+                    if letter in fsm.states:
+                        fsm.states[state].add_transition(letter, letter)
+                    else:
+                        fsm.states[state].add_transition(letter, current_state_name)
+            elif current_state_type == "accept":
+                for letter in automaton_alphabet:
+                    next = current_state_name*letter
+                    len_next_word = len(next)
+                    while True:
+                        if len(next) <= 1:
+                            if next in fsm.states:
+                                fsm.states[state].add_transition(letter, next)
+                            else:
+                                fsm.states[state].add_transition(letter, 'start')
+                            break
+                        else:
+                            if next in fsm.states:
+                                fsm.states[state].add_transition(letter, next)
+                                break
+                            next = next.subword(1, len(next))
+        return fsm
+
+    def reduce_using_automaton(self, word):
+        '''
+        The method for word reduction using automaton is mentioned in the section 13.1.3 of the Handbook.
+        All the elements of the automaton are stored in an array and are given as the input to the automaton.
+        If the automaton reaches a dead state that subword is replaced and the automaton is run from the beginning.
+        This is repeated until the word reaches the end and the automaton stays in the accept state.
+
+        '''
+        # compute all_rules.
+        all_rules = self.rules
+        self.compute_inverse_rules()
+        all_rules.update(self.inverse_rules)
+
+        flag = 1
+        while flag:
+            flag = 0
+            current_state = self.reduction_automaton.states['start']
+            word_array = [s for s in word.letter_form_elm]
+            for i in range (0, len(word_array)):
+                next_state_name = current_state.transitions[word_array[i]]
+                next_state = None
+                for state in self.reduction_automaton.states:
+                    if state == next_state_name:
+                        next_state = self.reduction_automaton.states[state]
+                if next_state.state_type == "dead":
+                    subst = all_rules[next_state_name]
```

Comment:
Perhaps, this informations could be stored in the reduction automaton. Maybe each state could have a `subst` attribute, containing the right hand side of the corresponding rule. It would be more efficient than computing all inverse rules again, when you already do this in `construct_automaton`.

Reply:
Have you thought about storing right hand sides of rules in the automaton? Then you wouldn't be recomputing `all_rules` in `reduce_using_automaton` (you already did it in `constuct_automaton`) and wouldn't need to have `inverse_rules` as an attribute of the `RewritingSystem` class.